### PR TITLE
Avoid out-of-bounds pointer arithmetic in dataLength()

### PR DESCRIPTION
**Pull request by DemiMarie**
This is undefined behavior.  Pinging @dmach @Conan-Kudo.

### CODE REVIEW

**Comment by Conan-Kudo** [comment]
It looks like some of the tests are failing on this?

**Comment by DemiMarie** [comment]
Not sure why test 346 failed, but I re-pushed it with just the bare minimal change.

**Review by Conan-Kudo — APPROVED** [review]

File: lib/header.c [hunk 1]
```diff
@@ -474,8 +474,8 @@ static int dataLength(rpm_tagtype_t type, rpm_constdata_t p, rpm_count_t count,
     default:
 	if (typeSizes[type] == -1)
 	    return -1;
-	length = typeSizes[(type & 0xf)] * count;
-	if (length < 0 || (se && (s + length) > se))
+	length = typeSizes[type] * count;
```
Comment (dmach) [inline lib/header.c:477]: The `& 0xf` was probably there to guarantee that we're accessing typeSizes within valid range.
Shouldn't we rather check the `type` range with the following? `if (type > RPM_MAX_TYPE) return -1;`

File: lib/header.c [hunk 2]
```diff
@@ -474,8 +474,8 @@ static int dataLength(rpm_tagtype_t type, rpm_constdata_t p, rpm_count_t count,
...
+	if (length < 0 || (se && length > se - s))
```
Comment (dmach) [inline lib/header.c:478]: Pointer arithmetic is sometimes tricky. If it's not a performance problem, I'd explicitly check that `se >= s`.

**Comment by pmatilai** [comment]
@DemiMarie I've repeatedly asked you not to submit more pull-requests of this kind, because a large percentage of these "but in theory" patches have only introduced regressions despite hours and days wasted trying to review them. Rpm relies on this type of arithmetic in any number of places, and AIUI it mostly just means that rpm is not portable to some truly exotic hardware with segmented memory model.

As I've also said before, revising the header/package parsing from the ant perspective isn't going to cut it.

**Comment by DemiMarie** [comment]
> @DemiMarie I've repeatedly asked you not to submit more pull-requests of this kind, because a large percentage of these "but in theory" patches have only introduced regressions despite hours and days wasted trying to review them. Rpm relies on this type of arithmetic in any number of places, and AIUI it mostly just means that rpm is not portable to some truly exotic hardware with segmented memory model.

It also can lead to miscompilations, which won’t be considered compiler bugs because this type of pointer arithmetic is undefined behavior.

**Pull request closed without merging**